source/dem/uniform_insertion.cc

```diff
@@ -63,24 +66,45 @@ UniformInsertion<dim>::insert(
         Utilities::MPI::all_gather(communicator, my_bounding_box);
 
       // Distbuting particles between processors
-      if (Utilities::MPI::this_mpi_process(communicator) !=
-          (Utilities::MPI::n_mpi_processes(communicator) - 1))
-        this->inserted_this_step_this_proc =
-          floor(this->inserted_this_step /
-                Utilities::MPI::n_mpi_processes(communicator));
-      else
+      this->inserted_this_step_this_proc =
+        floor(this->inserted_this_step / n_mpi_process);
+      if (this_mpi_process == (n_mpi_process - 1))
         this->inserted_this_step_this_proc =
           this->inserted_this_step -
-          (Utilities::MPI::n_mpi_processes(communicator) - 1) *
-            floor(this->inserted_this_step /
-                  Utilities::MPI::n_mpi_processes(communicator));
+          (n_mpi_process - 1) * floor(this->inserted_this_step / n_mpi_process);
+
+      Point<dim>              insertion_location;
+      std::vector<Point<dim>> insertion_points_on_proc;
+      insertion_points_on_proc.reserve(this->inserted_this_step_this_proc);
+
+      // Find insertion locations
+      if (this_mpi_process == (n_mpi_process - 1))
+        {
+          for (unsigned int id =
+                 this->inserted_this_step - this->inserted_this_step_this_proc;
+               id < this->inserted_this_step;
+               ++id)
+            {
+              find_insertion_location_uniform(insertion_location,
+                                              id,
+                                              dem_parameters.insertion_info);
+              insertion_points_on_proc.push_back(insertion_location);
+            }
+        }
+      else
+        {
+          for (unsigned int id =
+                 this_mpi_process * this->inserted_this_step_this_proc;
+               id < (this_mpi_process + 1) * this->inserted_this_step_this_proc;
+               ++id)
+            {
+              find_insertion_location_uniform(insertion_location,
+                                              id,
+                                              dem_parameters.insertion_info);
+              insertion_points_on_proc.push_back(insertion_location);
+            }
+        }
```

Comment:
I'm not sure this is ideal way to set this up. You could just set the variables for the start and the end of the loop using the if/else statement, then do the same loop for everyone and have them same function call. Would make understanding this easier.

Reply:
done.